# <center> 🐼 [Введение в Pandas](https://stepik.org/lesson/681964/)</center>

Этот ноутбук поможет разобраться вам в основных моментах предварительной обработки данных в Pandas, познакомиться с полезными функциями и трюками. 

### Оглавление ноутбука
<img src="../images/pandas.jpg" align="right" width="528" height="528" />
<br>

<p><font size="3" face="Arial" font-size="large">
<ul type="square"><li><a href="#1">📑 Структуры данных в Pandas</a></li><li><a href="#2">🔍 Разведочный анализ данных</a><ul><li><a href="#3">Загрузка и первичное знакомство</a></li><li><a href="#4">Предварительный осмотр</a></li><li><a href="#5">Работа с числовыми данными</a></li></ul></li><li><a href="#6">📤 Извлечение необходимой информации</a><ul><li><a href="#7">Сортировка</a></li><li><a href="#8">Методы фильтрации</a></li><li><a href="#9">Работа с пропусками</a></li><li><a href="#10">Индексация данных</a></li><li><a href="#11">Функции</a></li></ul></li>
</ul></font></p>

# <center>📑 Cтруктуры данных в Pandas. </center>

<p id="1"></p>

`Series` — это структура, используемая для работы с последовательностью одномерных данных, а `DataFrame` — более сложная и подходит для нескольких измерений. `Series` – одномерный массив с метками, способный хранить любой тип данных.

<img src="https://sun9-43.userapi.com/impg/0k7fzcgju-6hY10XmQoaJkfxOxr0-SvgA3NHQg/N0Hb-T_fXio.jpg?size=1280x720&quality=96&sign=6ac22e0ff168023b0dff780c1bfa5e36&type=album"></img>

In [43]:
import pandas as pd # pandas для работы с табличными данными
import numpy as np  # numpy для математических действий

### Создать`Series`  можно так:

`s = pd.Series(data, index=index, name='something')`

Метки именуются `index`, данные `data` могут быть любыми, `name` является именем `Series`.

In [44]:
data = ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog']
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
s = pd.Series(data, index=labels, name='animals')
s

a      cat
b      cat
c    snake
d      dog
e      dog
f      cat
g    snake
h      cat
i      dog
j      dog
Name: animals, dtype: str

### Создать `DataFrame` проще из словаря, а в качестве значений строк использовать список

In [45]:
data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
         'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
         'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
         'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

df = pd.DataFrame(data, index=labels)
df

,animal,age,visits,priority
a,cat,2.5,1,yes
b,cat,3.0,3,yes
c,snake,0.5,2,no
d,dog,NaN,3,yes
e,dog,5.0,2,no
f,cat,2.0,3,no
g,snake,4.5,1,no
h,cat,NaN,1,yes
i,dog,7.0,2,no
j,dog,3.0,1,no


### А теперь сделаем открытие! Столбцы в `DataFrame` являются `Series`!

In [46]:
s = df.age
s

a    2.5
b    3.0
c    0.5
d    NaN
e    5.0
f    2.0
g    4.5
h    NaN
i    7.0
j    3.0
Name: age, dtype: float64

Обращение по индексу `s[0]`

In [47]:
s.iloc[0]

np.float64(2.5)

Элементы по определенным индексам `s[[4, 3, 1]]`

In [48]:
s.iloc[[4, 3, 1]]

e    5.0
d    NaN
b    3.0
Name: age, dtype: float64

Срезы `s[:3]`



In [49]:
s.iloc[:3]

a    2.5
b    3.0
c    0.5
Name: age, dtype: float64

Идеология словаря при обращении к элементам датафрейма

In [50]:
s['a']

np.float64(2.5)

In [51]:
'e' in s

True

In [52]:
't' in s

False

__Арифметические операции к `Series`__



In [53]:
s + s

a     5.0
b     6.0
c     1.0
d     NaN
e    10.0
f     4.0
g     9.0
h     NaN
i    14.0
j     6.0
Name: age, dtype: float64

__Логические операции к `Series`.__ Например, выбрать все значения, больше заданного



In [54]:
s[s > 3]

e    5.0
g    4.5
i    7.0
Name: age, dtype: float64

Функции из модуля `numpy` применяются поэлементно. Например, экспонента `np.exp(s)`

Количество строк можно узнать так `len(s)`

##   Редактирование таблиц. Создание и удаление столбцов. 

<div class="alert alert-info">

К столбцам `DataFrame` применимы любые математические операции. Надо толькo понимать, как работает векторизация.

Так, мы можем присвоить всему столбцу `'visits'` значение 3 и оно автоматически появится в каждой строке:

In [55]:
df['visits'] = 3
df

,animal,age,visits,priority
a,cat,2.5,3,yes
b,cat,3.0,3,yes
c,snake,0.5,3,no
d,dog,NaN,3,yes
e,dog,5.0,3,no
f,cat,2.0,3,no
g,snake,4.5,3,no
h,cat,NaN,3,yes
i,dog,7.0,3,no
j,dog,3.0,3,no


Надо увеличить возраст на год?  Просто увеличиваем столбец `'age'` на единицу. Даже не верится, что все так просто! 



In [56]:
df.age = df.age + 1
df

,animal,age,visits,priority
a,cat,3.5,3,yes
b,cat,4.0,3,yes
c,snake,1.5,3,no
d,dog,NaN,3,yes
e,dog,6.0,3,no
f,cat,3.0,3,no
g,snake,5.5,3,no
h,cat,NaN,3,yes
i,dog,8.0,3,no
j,dog,4.0,3,no


Можем даже создать новый столбец, используя другие столбцы `'age'` и `'visits'`. Казалось бы, всего одна строчка, но применяется для всего столбца! Векторизация это секретная магия `pandas` =)

Аналогично, можно  вычесть, умножить, разделить и сложить два численных столбца друг с другом.

In [57]:
df['age_visits_sum'] = df.age + df.visits
df

,animal,age,visits,priority,age_visits_sum
a,cat,3.5,3,yes,6.5
b,cat,4.0,3,yes,7.0
c,snake,1.5,3,no,4.5
d,dog,NaN,3,yes,NaN
e,dog,6.0,3,no,9.0
f,cat,3.0,3,no,6.0
g,snake,5.5,3,no,8.5
h,cat,NaN,3,yes,NaN
i,dog,8.0,3,no,11.0
j,dog,4.0,3,no,7.0


<div class="alert alert-info">
    
<b>Важное замечание:</b>
При создании новой колонки рекомендуется писать её название без пробелов (хотя с ними тоже можно) - как в примере выше: `age_visits_sum`.
Тогда можно будет обращаться к ней не только как ключу словаря, но и как к параметру датафрейма через точку.<br>
**`df.age_visits_sum`**

In [58]:
# Например
df.age_visits_sum.value_counts()

age_visits_sum
7.0     2
6.5     1
4.5     1
9.0     1
6.0     1
8.5     1
11.0    1
Name: count, dtype: int64

Также можем и удалить столбец, не забываем про `inplace=True`.

In [59]:
df.drop('age_visits_sum', axis=1, inplace=True)
df

,animal,age,visits,priority
a,cat,3.5,3,yes
b,cat,4.0,3,yes
c,snake,1.5,3,no
d,dog,NaN,3,yes
e,dog,6.0,3,no
f,cat,3.0,3,no
g,snake,5.5,3,no
h,cat,NaN,3,yes
i,dog,8.0,3,no
j,dog,4.0,3,no


<div class="alert alert-info">
<b>Полезная информация</b>


- __Удаление столбцов или строк__ `df.drop(name, axis=0, inplace=True)`  - Тут `name` это индекс строки или название колонки. Параметр `axis=0` означает, что удаляется строка, для удаления колонок (столбцов) нужно писать вместо индекса имена колонок списком и ставить `axis=1`.

    
- __Удаление дубликатов строк__ `df.drop_duplicates()` и не забываем про параметр `inplace=True`. Если надо удалить дубли по конкретному столбцу, то передаем в параметр `subset` название или список названий.

    
    
- __Вставка в заданное место столбца__ `df.insert(loc, column, value)`.  Тут `loc` (куда), `column` (название), `value` (что вставлям). По умолчанию столбцы вставляются в конец.

    
    
- __Переименовать столбец__ можно так `df.rename(columns = {'old_name':'new_name'}, inplace=True)`</div>


# <center> 🔍 Разведочный анализ данных (EDA) </center>

<p id="2"></p>

В этой части непосредственно познакомимся с загрузкой данных, первичным осмотром и анализом.

 <img src="https://sun9-46.userapi.com/impg/4yA0_8N5ksJzftOvHsi_niXDlu4Ocy-HyjVj0w/YBBKjZhZsjc.jpg?size=1600x666&quality=96&sign=a6b10953218cb38c8e3042879b20935b&type=album"></img>

## Загрузка и первичное знакомство

<p id="3"></p>
<div class="alert alert-info">

**[pd.read_csv()](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)** - cчитывает данные в `DataFrame` из файла `csv` (файл с разделителями). У метода очень много параметров с которыми нужно ознакомиться самостоятельно.

In [60]:
path = '../data/demo_rides_info.csv'
rides_info = pd.read_csv(path)
rides_info.head(10)

,user_id,car_id,ride_id,ride_date,rating,ride_duration,ride_cost,speed_avg,speed_max,stop_times,distance,refueling,user_ride_quality,deviation_normal
0,q53527020Q,g55775125g,i1O,2020-01-16,3.42,33,324,75,100.298135,0,2.412860e+03,0,-3.828166,-12.577
1,X72952203K,Y-6420899e,Z1i,2020-01-19,4.00,65,839,51,78.000000,0,2.553284e+03,0,-0.166199,-5.850
2,g40994147T,z77468044J,x1M,2020-01-21,6.18,46,363,47,73.000000,2,2.079931e+02,0,-6.904072,0.000
3,E18650750E,W50413234I,W1Y,2020-03-17,1.43,80,955,34,116.000000,2,1.758479e+03,0,3.132158,18.966
4,D18083254C,e-4559210H,U1f,2020-03-21,3.45,142292,2276667,52,64.000000,1,7.541414e+06,0,-29.248641,-8.126
5,G12536963g,B61056692H,f1p,2020-02-26,5.73,49,681,36,65.000000,1,1.633555e+03,0,5.566720,0.000
6,j17924049w,g10190493K,e1F,2020-03-15,3.80,33,426,56,121.029522,0,1.854434e+03,0,-21.874094,-19.996
7,Z21861082q,U-8316679F,s1l,2020-02-22,2.88,23,685,53,58.000000,0,1.290301e+03,0,1.181858,-0.799
8,w32247420J,I-2277714Y,t1N,2020-02-08,0.10,168,8397,66,116.000000,2,5.049357e+03,0,-7.047228,16.841
9,i78201044i,T-1003767L,c1L,2020-01-02,0.00,72,930,75,146.456097,1,1.627029e+03,0,2.987268,0.000


<div class="alert alert-info"><b>Полезная информация</b>

- Если при чтении файла вы видите столцы с названием типа `Unnamed: 0` , то это сохраненный индекс при записи файла. Можно загрузить индексы из этого столбца, не создавая их по умолчанию с нуля, с помощью параметра  `index_col=0`, указав номер столбца который нужно использовать для индексов. 


- Полезные методы вывода строк: `df.head(n)` / `df.sample(n)` / `df.tail(n)` - возвращают первые, случайные или последние `n` строк датафрейма. Структура `DataFrame` А что это? `DataFrame` – это двумерная структура данных со столбцами потенциально разных типов.

    
- Подсчет числа встречаемости каждого значения:  `df.value_counts()` - функция используется для получения уникальных значений и числа их встречаемости в виде `Series` . По умолчанию отсортирован по убыванию встречаемости `(ascending=False)`. Можно изменить сортировку через `ascending=True`. Также числовые столбцы можно разбивать на бины через параметр `bins`.  

</div>

## Предварительный осмотр данных

<p id="4"></p>

**[df.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html)** - возвращает случайную выборку строк датафрейма. Параметр `n` - количество выводимых строк.

In [61]:
rides_info.sample(5)

,user_id,car_id,ride_id,ride_date,rating,ride_duration,ride_cost,speed_avg,speed_max,stop_times,distance,refueling,user_ride_quality,deviation_normal
25109,n16090039I,r-1341912M,A1x,2020-02-25,5.50,12,163,53,66.000000,0,457.790294,0,-27.640717,-3.651
49911,V54413806U,q-7310362e,D1D,2020-01-28,1.52,217,4335,32,119.000000,16,3133.402788,0,-8.734850,14.360
21813,u79574913Y,c48490818S,f1N,2020-02-05,7.52,43,682,47,84.656064,0,2066.290931,0,8.263095,-1.166
44092,e77586377t,e-8327999v,q1n,2020-02-17,5.45,56,890,45,77.000000,2,2233.880385,0,25.648136,41.535
48807,V12685254Z,l-1770973A,Y1K,2020-02-06,4.56,104,1658,43,66.000000,2,1802.015162,0,6.340388,-19.372


**[df.head()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html)** - возвращает первые строки датафрейма. По умолчанию выводит 5 строк, но можно изменить параметр `n` и вывести больше или меньше.  

Часто используют метод `head()`, но лучше первый раз смотреть на датафрейм через `sample()`.
Так вы имеете возможность смотреть на разные выборки и это может быть полезно, чтобы увидеть какие-нибудь баги.

In [62]:
rides_info.head()

,user_id,car_id,ride_id,ride_date,rating,ride_duration,ride_cost,speed_avg,speed_max,stop_times,distance,refueling,user_ride_quality,deviation_normal
0,q53527020Q,g55775125g,i1O,2020-01-16,3.42,33,324,75,100.298135,0,2.412860e+03,0,-3.828166,-12.577
1,X72952203K,Y-6420899e,Z1i,2020-01-19,4.00,65,839,51,78.000000,0,2.553284e+03,0,-0.166199,-5.850
2,g40994147T,z77468044J,x1M,2020-01-21,6.18,46,363,47,73.000000,2,2.079931e+02,0,-6.904072,0.000
3,E18650750E,W50413234I,W1Y,2020-03-17,1.43,80,955,34,116.000000,2,1.758479e+03,0,3.132158,18.966
4,D18083254C,e-4559210H,U1f,2020-03-21,3.45,142292,2276667,52,64.000000,1,7.541414e+06,0,-29.248641,-8.126


<div class="alert alert-info">
    <b>Выводы</b>
    <ul>
    <li>в целом данные загрузились</li>
    <li>названия столбцов из первой строки загружены</li>
    <li>индексы первого столбца выглядят адекватно</li>
    <li>наличие индексов, которые нужно удалить, не обнаружено</li>
    <li>по случайным пяти строкам наличие каких-либо ошибок не обнаружено</li></ul>
</div>

<div class="alert alert-info">

**[df.info()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html)*** - Выводит краткую сводку-информацию по датафрейму.

В нашем случае датафрейм небольшой, но он может быть большим. Мы разберем методы библиотеки `Pandas` на этом небольшом примере, которые вы по аналогии сможете самостоятельно применить к другим датасетам.

In [63]:
rides_info.info()

<class 'pandas.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   user_id            50000 non-null  str    
 1   car_id             50000 non-null  str    
 2   ride_id            50000 non-null  str    
 3   ride_date          50000 non-null  str    
 4   rating             50000 non-null  float64
 5   ride_duration      50000 non-null  int64  
 6   ride_cost          50000 non-null  int64  
 7   speed_avg          50000 non-null  int64  
 8   speed_max          49747 non-null  float64
 9   stop_times         50000 non-null  int64  
 10  distance           50000 non-null  float64
 11  refueling          50000 non-null  int64  
 12  user_ride_quality  49841 non-null  float64
 13  deviation_normal   50000 non-null  float64
dtypes: float64(5), int64(5), str(4)
memory usage: 5.3 MB


<div class="alert alert-info"><b>Выводы</b>

* в датафрейме всего 14 столбцов
* четыре столбца имеют тип `object`
* пять столбцов имеют тип `float64`
* пять столбцов имеют тип `int64`
* всего в датафрейме 50000 строк
* в столбцах `speed_max` и `user_ride_quality` есть пропуски (nan)
* объем занимаемый датафреймом `5.3+ MB` (эта информация может быть полезна, чтобы заранее представить какие операции с датафреймом вы сможете выполнять с учетом доступной вам оперативной памяти)


<div class="alert alert-info"><b>Выводы</b>

**[df.columns](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.columns.html)*** - Выводит метки столбцов датафрейма.  

<div class="alert alert-info"><b>Выводы:</b>
    
Иногда количество столбцов очень большое, так что при выводе они все не помещаются в стандартный вывод через `df.info()`. Тогда можно воспользоваться атрибутом для просмотра индексов названий столбцов.

In [64]:
rides_info.columns

Index(['user_id', 'car_id', 'ride_id', 'ride_date', 'rating', 'ride_duration',
       'ride_cost', 'speed_avg', 'speed_max', 'stop_times', 'distance',
       'refueling', 'user_ride_quality', 'deviation_normal'],
      dtype='str')

Гораздо нагляднее и удобнее работать со списком названий столбцов

In [65]:
column_names = list(rides_info.columns)
column_names, len(column_names)

(['user_id',
  'car_id',
  'ride_id',
  'ride_date',
  'rating',
  'ride_duration',
  'ride_cost',
  'speed_avg',
  'speed_max',
  'stop_times',
  'distance',
  'refueling',
  'user_ride_quality',
  'deviation_normal'],
 14)

<div class="alert alert-info"><b>Выводы</b>
<ul><li>14 столбцов</li>
<li>все названия столбцов в нижнем регистре и не содержат спецсимволов (значит их не нужно переименовывать, чтобы обращаться к ним через точку)</li></ul></div>


## Работа с числовыми данными

<div class="alert alert-info">
<p id="5"></p>

**[df.describe()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html)*** - Выводит описательные статистики по числовым столбцам датафрейма. Описательные статистики включают в себя
- `count` - количество не пустых значений
- `mean` - среднее арифметическое
- `std` - стандартное отклонение
- `min`, `max` - минимум и максимум
- 25%, 50%, 75% - соотвествующие квартили  


__!!!Bнимание!!!__ из статистик автоматически исключаются `NaN` значения, если они есть.  

Выведем такое описание таблицы, округлив значения до второго знака после запятой:

In [66]:
rides_info.describe().round(2)

,rating,ride_duration,ride_cost,speed_avg,speed_max,stop_times,distance,refueling,user_ride_quality,deviation_normal
count,50000.00,50000.00,50000.00,50000.00,49747.00,50000.00,50000.00,50000.0,49841.00,50000.00
mean,4.49,9304.08,117999.44,47.10,77.48,1.34,444872.13,0.0,-0.19,-1.20
std,2.13,35947.59,504330.21,12.65,24.28,2.35,1782500.58,0.0,11.79,19.40
min,0.00,2.00,9.00,25.00,30.00,0.00,3.04,0.0,-59.52,-110.76
25%,3.14,27.00,297.00,38.00,62.00,0.00,784.16,0.0,-6.94,-9.01
50%,4.49,44.00,506.00,46.00,72.00,1.00,1456.72,0.0,0.16,0.00
75%,5.86,69.00,895.00,52.00,86.00,2.00,2265.27,0.0,6.69,6.98
max,10.00,222356.00,10889545.00,100.00,202.46,21.00,18333880.69,1.0,56.53,99.92


<div class="alert alert-info">
    <b>Выводы</b>
    
- описание столбцов датафрейма (`speed_avg` - средняя скорость по маршруту такси, `rating` - рейтинг, полученный водителем за поездку, `ride_duration` - продолжительность поездки, `ride_cost` - стоимость поездки, `stop_times` - количество остановок)
- средняя скорость движения по маршруту по всем поездкам около `47`, а максимальная средняя скорость `100`. Вероятно, этот столбец имеет размерность км/час
- как минимум в половине случаев во время поездки была одна остановка
- максимальный рейтинг около `10`, минимум `0`. Более половины всех рейтингов с оценками `от 3 до 5`. Вероятно, размерность этого столбца - баллы
- сделать однозначные выводы о размерностях продолжительности и стоимости поездок нельзя, вероятно, они были модифицированы и для этого требуется дополнительный анализ и визуализация данных
</div>

`describe()` также можно использовать и для категориальных столбцов, добавив параметр `include='object'`

In [67]:
rides_info.describe(include='object')

C:\Users\b_jumaev\AppData\Local\Temp\ipykernel_16744\3686109808.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  rides_info.describe(include='object')


,user_id,car_id,ride_id,ride_date
count,50000,50000,50000,50000
unique,13975,4250,2704,92
top,q75765658l,o-1829013e,H1v,2020-01-16
freq,18,25,34,604


**[df.value_counts()](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html)** - функция используется для получения уникальных значений и числа их встречаемости в виде `Series`. По умолчанию отсортирован по убыванию встречаемости (`ascending=False`). Можно изменить сортировку через `ascending=True`. Также числовые столбцы можно разбивать на бины через параметр `bins`.

In [68]:
rides_info.ride_id.value_counts()[:5]

ride_id
H1v    34
E1y    34
P1Q    33
K1F    33
H1z    33
Name: count, dtype: int64

Вывод: самый часто встречающийся код маршрута `H1v` и `E1y`, имеют по 34 поездки

Посмотрим как распределяется кол-во поездок по рейтингу поездок разбивая его на бины:

In [69]:
rides_info.rating.value_counts(bins=4)

(2.5, 5.0]       21730
(5.0, 7.5]       15101
(-0.011, 2.5]     8761
(7.5, 10.0]       4408
Name: count, dtype: int64

Вывод: 26% (8761+4408 делить на 50000) поездок с рейтингом ниже 2.5 и выше 7.5



**[.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html)** - метод, который позволит нам выявить дублирование в данных. Возвращает булевы значения, обозначающие повторяющиеся строки. При этом можно не учитывать некоторые определенные столбцы.

In [70]:
df = pd.DataFrame({
    'brand': ['Yum Yum', 'Yum Yum', 'Indomie', 'Indomie', 'Indomie'],
    'style': ['cup', 'cup', 'cup', 'pack', 'pack'],
    'rating': [4, 4, 3.5, 15, 5]
})
df.duplicated()

0    False
1     True
2    False
3    False
4    False
dtype: bool

Также можно использовать на определенном столбце/столбцах.

In [71]:
df.duplicated(subset=['brand'])

0    False
1     True
2    False
3     True
4     True
dtype: bool

# <center> 📤 Извлечение необходимой информации </center>

<p id="6">Перейдем к самому интересному - непосредственно к анализу полученных данных, познакомимся с методами сортировки, фильтрации и разными функциями.</p>

<img src="https://sun9-41.userapi.com/impg/io095r0KS1Fq4NTHYReNuRTf6DBM1pihDZMPqQ/Xki0IFcFAK8.jpg?size=2536x1150&quality=96&sign=db8f9c09f39dd1417032d06830270d44&type=album"></img>

## Сортировка `DataFrame`

<div class="alert alert-info">
<p id="7">
    
`DataFrame` можно сортировать. Посмотрим как это можно делать.</p>


In [72]:
data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
         'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
         'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
         'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

df = pd.DataFrame(data, index=labels)
df

,animal,age,visits,priority
a,cat,2.5,1,yes
b,cat,3.0,3,yes
c,snake,0.5,2,no
d,dog,NaN,3,yes
e,dog,5.0,2,no
f,cat,2.0,3,no
g,snake,4.5,1,no
h,cat,NaN,1,yes
i,dog,7.0,2,no
j,dog,3.0,1,no


__По значениям__

In [73]:
df.sort_values(by='age') 

,animal,age,visits,priority
c,snake,0.5,2,no
f,cat,2.0,3,no
a,cat,2.5,1,yes
b,cat,3.0,3,yes
j,dog,3.0,1,no
g,snake,4.5,1,no
e,dog,5.0,2,no
i,dog,7.0,2,no
d,dog,NaN,3,yes
h,cat,NaN,1,yes


__По индексу__

In [74]:
df.sort_index(axis=1, ascending=False)

,visits,priority,animal,age
a,1,yes,cat,2.5
b,3,yes,cat,3.0
c,2,no,snake,0.5
d,3,yes,dog,NaN
e,2,no,dog,5.0
f,3,no,cat,2.0
g,1,no,snake,4.5
h,1,yes,cat,NaN
i,2,no,dog,7.0
j,1,no,dog,3.0


<div class="alert alert-info">
    
Параметр `ascending=True` по умолчанию, соответственно сортируется по возрастанию. Если нужно сортировать по нескольким столбцам, то передается список столбов и список булевых переменных в `ascending`. Параметр `inplace=False` по умолчанию.



## Фильтрация данных

<p id="8">Логические выражения используются для получения подтаблицы с заданными условиями по столбцам.</p>

Рассмотрим примеры.



In [75]:
df.loc[df.visits > 2]

,animal,age,visits,priority
b,cat,3.0,3,yes
d,dog,NaN,3,yes
f,cat,2.0,3,no


Если условий больше одного, то условия фильтрации записываются в круглых скобках!

- __'И'__ обозначается как __&__
- __'ИЛИ'__  обозначается как __|__

In [76]:
df.loc[(df.visits > 2) & (df.priority == 'yes')]

,animal,age,visits,priority
b,cat,3.0,3,yes
d,dog,NaN,3,yes


In [77]:
df.loc[(df.visits > 2) | (df.age > 5)]

,animal,age,visits,priority
b,cat,3.0,3,yes
d,dog,NaN,3,yes
f,cat,2.0,3,no
i,dog,7.0,2,no


Чтобы нам вывести часть датафрейма с дешевыми ценами (цены поездки ниже 500), попробуем просто записать это условие и вывести часть датафрейма, которая соответствует этому условию 

In [78]:
rule_cheap = rides_info.ride_cost < 500
rides_info[rule_cheap].head().round(2)

,user_id,car_id,ride_id,ride_date,rating,ride_duration,ride_cost,speed_avg,speed_max,stop_times,distance,refueling,user_ride_quality,deviation_normal
0,q53527020Q,g55775125g,i1O,2020-01-16,3.42,33,324,75,100.30,0,2412.86,0,-3.83,-12.58
2,g40994147T,z77468044J,x1M,2020-01-21,6.18,46,363,47,73.00,2,207.99,0,-6.90,0.00
6,j17924049w,g10190493K,e1F,2020-03-15,3.80,33,426,56,121.03,0,1854.43,0,-21.87,-20.00
11,N20390973B,T12315854s,g1g,2020-03-18,8.59,27,372,48,66.00,1,1061.04,0,-9.69,-18.86
12,x37836441s,o-5998359x,W1M,2020-01-05,7.88,14,177,38,46.00,2,470.41,0,7.31,2.39


### Булева маска

Также один из самых простых способов отфильтровать нужные строки - это добавить булевую маску. Пусть маска строится по условию, что стоимость поездки больше некоторого значения.

- Создаем булеву маску  `[True, False, True, ...]`
- Выводим первые 5 значений строк, подходящих под условие

In [79]:
rule_cheap = rides_info['ride_cost'] < 426  
rides_info[rule_cheap].head()   

,user_id,car_id,ride_id,ride_date,rating,ride_duration,ride_cost,speed_avg,speed_max,stop_times,distance,refueling,user_ride_quality,deviation_normal
0,q53527020Q,g55775125g,i1O,2020-01-16,3.42,33,324,75,100.298135,0,2412.859858,0,-3.828166,-12.577
2,g40994147T,z77468044J,x1M,2020-01-21,6.18,46,363,47,73.000000,2,207.993094,0,-6.904072,0.000
11,N20390973B,T12315854s,g1g,2020-03-18,8.59,27,372,48,66.000000,1,1061.035898,0,-9.692375,-18.856
12,x37836441s,o-5998359x,W1M,2020-01-05,7.88,14,177,38,46.000000,2,470.414162,0,7.314068,2.389
16,v23018580U,J-9977061N,e1v,2020-02-18,5.01,30,415,35,80.000000,0,1045.715427,0,-5.538496,18.364


Если мы хотим добавить новый столбец `cheap_price` - характеристика дешевой цены поездки: если цена в первом 25% квартиле (то есть меньше 426) то это будет дешевая цена, поставим значение `Yes`, иначе `Nan`.

In [80]:
rule_cheap = rides_info.ride_cost < 426
rides_info.loc[rule_cheap, 'cheap_price'] = 'Yes'
rides_info.head().round(2)

,user_id,car_id,ride_id,ride_date,rating,ride_duration,ride_cost,speed_avg,speed_max,stop_times,distance,refueling,user_ride_quality,deviation_normal,cheap_price
0,q53527020Q,g55775125g,i1O,2020-01-16,3.42,33,324,75,100.3,0,2412.86,0,-3.83,-12.58,Yes
1,X72952203K,Y-6420899e,Z1i,2020-01-19,4.00,65,839,51,78.0,0,2553.28,0,-0.17,-5.85,NaN
2,g40994147T,z77468044J,x1M,2020-01-21,6.18,46,363,47,73.0,2,207.99,0,-6.90,0.00,Yes
3,E18650750E,W50413234I,W1Y,2020-03-17,1.43,80,955,34,116.0,2,1758.48,0,3.13,18.97,NaN
4,D18083254C,e-4559210H,U1f,2020-03-21,3.45,142292,2276667,52,64.0,1,7541414.25,0,-29.25,-8.13,NaN


### Методы `.filter()` и `.query()`


Выбор строк или столбцов



In [81]:
df.filter(items=['a', 'j'], axis=0)

,animal,age,visits,priority
a,cat,2.5,1,yes
j,dog,3.0,1,no


<div class="alert alert-info">

__Пояснение__

Этот метод также имеет параметр `like`. Он отбирает строки / столбцы, в названии которых есть подстрока, присвоенная ему. Параметры `items` и `like` взаимоисключающие.



In [82]:
df.filter(like='a', axis=1)

,animal,age
a,cat,2.5
b,cat,3.0
c,snake,0.5
d,dog,NaN
e,dog,5.0
f,cat,2.0
g,snake,4.5
h,cat,NaN
i,dog,7.0
j,dog,3.0


__Фильтрация__ альтернативным способом



In [83]:
df.query('age > visits')

,animal,age,visits,priority
a,cat,2.5,1,yes
e,dog,5.0,2,no
g,snake,4.5,1,no
i,dog,7.0,2,no
j,dog,3.0,1,no


<div class="alert alert-info">

__Пояснение__

Если в запросе необходимо использовать переменную, например, `mean_age` то перед ней пишем `@`

`df.query('age > @mean_age')`

Можно использовать сложные запросы с __'И'__ обозначается как `&`, __'ИЛИ'__  обозначается как `|`.

`df.query('age > @mean_age | animal == 'cat')`




### Метод [`concat()`](https://pandas.pydata.org/docs/reference/api/pandas.concat.html) позволяет объединять объекты pandas вдоль определенной оси.

In [84]:
s1 = pd.Series(['a', 'b'])
s2 = pd.Series(['c', 'd'])
pd.concat([s1, s2])

0    a
1    b
0    c
1    d
dtype: str

Также есть возможность задавать индексы и лэйблы.

In [85]:
pd.concat([s1, s2], keys=['s1', 's2'],
          names=['Series name', 'Row ID'])

Series name  Row ID
s1           0         a
             1         b
s2           0         c
             1         d
dtype: str

Пример соединения `DataFrame`.

In [86]:
df1 = pd.DataFrame([['a', 1], ['b', 2]],
                   columns=['letter', 'number'])
df2 = pd.DataFrame([['c', 3], ['d', 4]],
                   columns=['letter', 'number'])

pd.concat([df1, df2])

,letter,number
0,a,1
1,b,2
0,c,3
1,d,4


### __Метод `isin()`__ - это один из самых полезных способов создать булевую маску для фильтрации

In [87]:
car2stay = ['g55775125g', 'Y-6420899e', 'z77468044J']
mask = rides_info['car_id'].isin(car2stay)

In [88]:
rides_info[mask].head()

,user_id,car_id,ride_id,ride_date,rating,ride_duration,ride_cost,speed_avg,speed_max,stop_times,distance,refueling,user_ride_quality,deviation_normal,cheap_price
0,q53527020Q,g55775125g,i1O,2020-01-16,3.42,33,324,75,100.298135,0,2.412860e+03,0,-3.828166,-12.577,Yes
1,X72952203K,Y-6420899e,Z1i,2020-01-19,4.00,65,839,51,78.000000,0,2.553284e+03,0,-0.166199,-5.850,NaN
2,g40994147T,z77468044J,x1M,2020-01-21,6.18,46,363,47,73.000000,2,2.079931e+02,0,-6.904072,0.000,Yes
5564,Y10710002e,Y-6420899e,K1s,2020-02-28,1.70,135,1884,64,66.336795,0,3.233434e+03,0,9.213972,-52.169,NaN
8545,K18258387y,g55775125g,Q1N,2020-02-08,4.49,60889,852440,57,66.920236,0,3.587582e+06,0,-15.355805,0.413,NaN


In [89]:
rides_info[~mask].head()

,user_id,car_id,ride_id,ride_date,rating,ride_duration,ride_cost,speed_avg,speed_max,stop_times,distance,refueling,user_ride_quality,deviation_normal,cheap_price
3,E18650750E,W50413234I,W1Y,2020-03-17,1.43,80,955,34,116.000000,2,1.758479e+03,0,3.132158,18.966,NaN
4,D18083254C,e-4559210H,U1f,2020-03-21,3.45,142292,2276667,52,64.000000,1,7.541414e+06,0,-29.248641,-8.126,NaN
5,G12536963g,B61056692H,f1p,2020-02-26,5.73,49,681,36,65.000000,1,1.633555e+03,0,5.566720,0.000,NaN
6,j17924049w,g10190493K,e1F,2020-03-15,3.80,33,426,56,121.029522,0,1.854434e+03,0,-21.874094,-19.996,NaN
7,Z21861082q,U-8316679F,s1l,2020-02-22,2.88,23,685,53,58.000000,0,1.290301e+03,0,1.181858,-0.799,NaN


## Работа с пропущенными значениями

<p id="9"></p>

Pandas для представления отсутствующих данных использует значение из numpy `np.nan`. По умолчанию оно не включается в вычисления. Где хранятся `np.nan`  показывает `df.isna()` оно же `df.isnull()`



In [90]:
df.isnull()

,animal,age,visits,priority
a,False,False,False,False
b,False,False,False,False
c,False,False,False,False
d,False,True,False,False
e,False,False,False,False
f,False,False,False,False
g,False,False,False,False
h,False,True,False,False
i,False,False,False,False
j,False,False,False,False


Метод `dropna()` удаляет NaN's из датасета

In [91]:
df.dropna()

,animal,age,visits,priority
a,cat,2.5,1,yes
b,cat,3.0,3,yes
c,snake,0.5,2,no
e,dog,5.0,2,no
f,cat,2.0,3,no
g,snake,4.5,1,no
i,dog,7.0,2,no
j,dog,3.0,1,no


<div class="alert alert-info">

__Замечание__

По умолчанию некоторые параметры - этого метода следующие:

- `axis=0`  (`0` удаляет строки, содержащие пропущенные значения, `1` удаляет столбцы)
- `how='any'` (`'any'` если хотя бы одно `NA`, удаляет эту строку или столбец, `'all'` только если все значения `NA`)
- `inplace=False` (`False` выполняет операцию и возвращает копию объекта, True выполняет операцию на месте и возвращает `None`)


Вжух, и все пропущенные значения исчезли! Однако, не забывайте про параметр `inplace=True`, иначе ваши изменения не сохранятся.

Заполняет пропущенные значения `df.fillna(n)`, при `n = 5` имеем:

In [92]:
df.fillna(5)

,animal,age,visits,priority
a,cat,2.5,1,yes
b,cat,3.0,3,yes
c,snake,0.5,2,no
d,dog,5.0,3,yes
e,dog,5.0,2,no
f,cat,2.0,3,no
g,snake,4.5,1,no
h,cat,5.0,1,yes
i,dog,7.0,2,no
j,dog,3.0,1,no


<div class="alert alert-info">
    
По умолчанию некоторые параметры этого метода следующие

- `axis=None` (`0` заполняет по строкам, `1` заполняет по столбцам)

- `inplace=False` (`False` выполняет операцию и возвращает копию объекта, `True` выполняет операцию на месте и возвращает `None`)

Количество пропусков в каждом столбце

In [93]:
df.isna().sum()

animal      0
age         2
visits      0
priority    0
dtype: int64

`df.fillna()` - Заполняет значения `NA/NaN`. Параметр `inplace=True` заполняет в датафрейме без необходимости копирования. По умолчанию, `inplace=False`.

## Основы индексации данных


<p id="10"></p>

**[df.loc[]](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html)** - Обеспечивает доступ к группе строк и столбцов по меткам (индексам) или логическому массиву. Обратите внимание, что, в отличие от обычных срезов `Python`, включены как начало, так и конец индексов. Через этот метод можно менять значения датафрейма.

Также, выбрать столбец можно как `df.col`, где `col` - это имя столбца, если имя столбца написано латиницей и не содержит пробелов. На примерах это будет выглядеть так:

In [94]:
df.animal

a      cat
b      cat
c    snake
d      dog
e      dog
f      cat
g    snake
h      cat
i      dog
j      dog
Name: animal, dtype: str

In [95]:
df.loc['a', ['animal', 'visits']]

animal    cat
visits      1
Name: a, dtype: object

In [96]:
df.iloc[:,0]

a      cat
b      cat
c    snake
d      dog
e      dog
f      cat
g    snake
h      cat
i      dog
j      dog
Name: animal, dtype: str

In [97]:
df.iloc[[1, 3], [0]]

,animal
b,cat
d,dog


<div class="alert alert-info">

Наглядно это выглядит так

<img src="https://sun9-79.userapi.com/impg/p0NrZ2Owypd4NbKQV4--0kZYi4PfNqk3E31URQ/7NTTharc5XA.jpg?size=301x129&quality=96&sign=36a4edc1d86f1b6618a80545f36cf4ef&type=album"></img>

__Разница между `iloc` и `loc` простая!__   `.loc`- возвращает строку по номеру, `.iloc` по индексу. <br>
**Важно:** Ещё одно различие состоит в том, что при наличии числового индекса `.iloc()` не будет включать правую границу в срез, а `.loc()` будет.

Посмотрим на примере:

In [98]:
# Строка с индексом 3 и колонка 2 не включены в срез
df.iloc[1:3, 0:2]

,animal,age
b,cat,3.0
c,snake,0.5


In [99]:
# А здесь правые границы включены в срез
df.reset_index().loc[1:3, 'animal':'visits']

,animal,age,visits
1,cat,3.0,3
2,snake,0.5,2
3,dog,NaN,3


### Фокусы с индексами


Сбрасывает индекс `df.reset_index()`, причем предыдущий индекс не удаляется, а сохраняется в новой колонке. <br>

In [100]:
df.reset_index()

,index,animal,age,visits,priority
0,a,cat,2.5,1,yes
1,b,cat,3.0,3,yes
2,c,snake,0.5,2,no
3,d,dog,NaN,3,yes
4,e,dog,5.0,2,no
5,f,cat,2.0,3,no
6,g,snake,4.5,1,no
7,h,cat,NaN,1,yes
8,i,dog,7.0,2,no
9,j,dog,3.0,1,no


Чтобы удалить предыдущий индекс и сохранить изменения, нужно указать параметры `drop=True, inplace=True`. 

In [101]:
df2 = df.reset_index(drop=True)
df2.head()

,animal,age,visits,priority
0,cat,2.5,1,yes
1,cat,3.0,3,yes
2,snake,0.5,2,no
3,dog,NaN,3,yes
4,dog,5.0,2,no


Устанавливает индекс `df.set_index()`. Соответственно, берет существующий столбец и переводит его в индекс.



Заменяет индекс `df.reindex()`



In [102]:
df.reindex(['j', 'i', 'h', 0, 1, 2], axis=0, fill_value=5)

TypeError: Invalid value '5' for dtype 'str'. Value should be a string or missing value, got 'int' instead.

<p id="11"></p>

## Функции
Применение функции к данным по строкам/столбцам реализуется как `df.apply(func, axis=0)`

<img src="https://sun9-70.userapi.com/impg/SfdR2m0D2E02P1xeDa8XJTr2DopfykEufD7Kcw/rQkknfzl800.jpg?size=216x398&quality=96&sign=b395742fb9db51dbeb0e124d17f82901&type=album"></img>

В примере методу `apply()` передается функция, которая будет применяться к данным. Стоит заметить, что значения типа `str` тоже суммируются.

In [ ]:
df.apply(np.sum, axis=0)

animal      catcatsnakedogdogcatsnakecatdogdog
age                                       27.5
visits                                      19
priority              yesyesnoyesnononoyesnono
dtype: object

Применение функции к данным поэлементно реализуется как `df.applymap(lambda x: func)`

В примере методу `applymap()`  передается `lambda`-функция, которая будет применяться к элементу.

In [ ]:
df.applymap(lambda x: len(str(x)))

,animal,age,visits,priority
a,3,3,1,3
b,3,3,1,3
c,5,3,1,2
d,3,3,1,3
e,3,3,1,2
f,3,3,1,2
g,5,3,1,2
h,3,3,1,3
i,3,3,1,2
j,3,3,1,2


<div class="alert alert-info">

Также функцию `apply()` можно применять для создания новых колонок на основе уже существующей или на основе всего датафрейма. <br>
При этом есть существенная разница при обращении к одной колонке или ко всему датасету.<br>
Разберём на примере:

In [ ]:
df.head()

,animal,age,visits,priority
a,cat,2.5,1,yes
b,cat,3.0,3,yes
c,snake,0.5,2,no
d,dog,NaN,3,yes
e,dog,5.0,2,no


In [ ]:
# Создадим бинарную колонку 'more_2', если было больше 2-х визитов на основе колонки 'visits'
df['more_2'] = df['visits'].apply(lambda x: 1 if x > 2 else 0)
df.head()

,animal,age,visits,priority,more_2
a,cat,2.5,1,yes,0
b,cat,3.0,3,yes,1
c,snake,0.5,2,no,0
d,dog,NaN,3,yes,1
e,dog,5.0,2,no,0


In [ ]:
# Теперь создадим такую же колонку с названием 'more_2v2' на основе всего df
df['more_2v2'] = df.apply(lambda x: 1 if x.visits > 2 else 0, axis=1)
df.head()

,animal,age,visits,priority,more_2,more_2v2
a,cat,2.5,1,yes,0,0
b,cat,3.0,3,yes,1,1
c,snake,0.5,2,no,0,0
d,dog,NaN,3,yes,1,1
e,dog,5.0,2,no,0,0


Отметим, что при использовании `apply()` к датасету - надо указывать название колонки через точку и параметр `axis`.

### Встроенные функции

Здесь перечислены некоторые полезные методы, применимые к `Series`

<img src="https://sun9-17.userapi.com/impg/QIvfQYVUFnOii25XfA6yL9HSsnVaMSI71Ccqow/BROCketQrDw.jpg?size=1302x1068&quality=96&sign=ffe48fa08279d37d7e023cce818f888c&type=album" width=700></img>

Например, посчитаем количество уникальных животных:

In [ ]:
df.animal.nunique()

3

<div class="alert alert-info">
    
<b> Заключение: </b>
В этом уроке мы познакомились с основными возможностями библиотеки `Pandas`, для закрепления материала рекомендуем прорешать задачи в степах.<br>
С более продвинутыми функциями библиотеки и их применением для решения практических задач будем разбираться в следующих уроках.